code/modules/modular_computers/file_system/programs/research_app.dm

```diff
@@ -0,0 +1,218 @@
+				return TRUE
+			research_node(params["node_id"], usr)
+			return TRUE
+
+/datum/computer_file/program/science/ui_static_data(mob/user)
+	. = list(
+		"static_data" = list()
+	)
+
+	// Build node cache...
+	// Note this looks a bit ugly but its to reduce the size of the JSON payload
+	// by the greatest amount that we can, as larger JSON payloads result in
+	// hanging when the user opens the UI
+	var/node_cache = list()
+	for (var/nid in SSresearch.techweb_nodes)
+		var/datum/techweb_node/n = SSresearch.techweb_nodes[nid] || SSresearch.error_node
+		var/cid = "[compress_id(n.id)]"
+		node_cache[cid] = list(
+			"name" = n.display_name,
+			"description" = n.description
+		)
+		if (n.research_costs?.len)
+			node_cache[cid]["costs"] = list()
+			for (var/c in n.research_costs)
+				node_cache[cid]["costs"]["[compress_id(c)]"] = n.research_costs[c]
+		if (n.prereq_ids?.len)
+			node_cache[cid]["prereq_ids"] = list()
+			for (var/pn in n.prereq_ids)
+				node_cache[cid]["prereq_ids"] += compress_id(pn)
+		if (n.design_ids?.len)
+			node_cache[cid]["design_ids"] = list()
+			for (var/d in n.design_ids)
+				node_cache[cid]["design_ids"] += compress_id(d)
+		if (n.unlock_ids?.len)
+			node_cache[cid]["unlock_ids"] = list()
+			for (var/un in n.unlock_ids)
+				node_cache[cid]["unlock_ids"] += compress_id(un)
+		if (n.required_experiments?.len)
+			node_cache[cid]["required_experiments"] = n.required_experiments
+		if (n.discount_experiments?.len)
+			node_cache[cid]["discount_experiments"] = n.discount_experiments
+
+	// Build design cache
+	var/design_cache = list()
+	var/datum/asset/spritesheet/research_designs/ss = get_asset_datum(/datum/asset/spritesheet/research_designs)
+	var/size32x32 = "[ss.name]32x32"
+	for (var/did in SSresearch.techweb_designs)
+		var/datum/design/d = SSresearch.techweb_designs[did] || SSresearch.error_design
+		var/cid = "[compress_id(d.id)]"
+		var/size = ss.icon_size_id(d.id)
+		design_cache[cid] = list(
+			d.name,
+			"[size == size32x32 ? "" : "[size] "][d.id]"
+		)
+
+	// Ensure id cache is included for decompression
+	var/flat_id_cache = list()
+	for (var/id in id_cache)
+		flat_id_cache += id
+
+	.["static_data"] = list(
+		"node_cache" = node_cache,
+		"design_cache" = design_cache,
+		"id_cache" = flat_id_cache
+	)
+
+/**
+ * Compresses an ID to an integer representation using the id_cache, used for deduplication
+ * in sent JSON payloads
+ *
+ * Arguments:
+ * * id - the ID to compress
+ */
+/datum/computer_file/program/science/proc/compress_id(id)
+	if (!id_cache[id])
+		id_cache[id] = id_cache_seq++
+	return id_cache[id]
+
+/datum/computer_file/program/science/proc/research_node(id, mob/user)
+	if(!stored_research.available_nodes[id] || stored_research.researched_nodes[id])
+		computer.say("Node unlock failed: Either already researched or not available!")
+		return FALSE
+	var/datum/techweb_node/TN = SSresearch.techweb_node_by_id(id)
+	if(!istype(TN))
```

Comment:
```suggestion
	var/datum/techweb_node/node = SSresearch.techweb_node_by_id(id)
	if(!istype(node))
```
This line does not meet our contribution standards, specifically for [variable name compliance](https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names), and should be refactored.